### PR TITLE
Separate storage code to storage pkg - continue

### PR DESCRIPTION
**Pull request by ShellyKa13**
**What this PR does / why we need it**:
Continue with moving storage related code to storage pkg.
This PR focuses on dv  and pvc related code.

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/storage/types/pvc.go [hunk 1]
```diff
@@ -155,3 +157,45 @@ func GetSizeIncludingFSOverhead(size *resource.Quantity, storageClass *string, v
...
+	}
+
+	pvc, ok := obj.(*k8sv1.PersistentVolumeClaim)
+	if !ok {
+		return nil, fmt.Errorf("error converting object to PersistentVolumeClaim: object is of type %T", obj)
+	}
+
+	return pvc, nil
+}
+func HasPVCBinding(namespace string, volumes []virtv1.Volume, pvcInformer cache.SharedIndexInformer) bool {
```
Comment (brybacki) [inline pkg/storage/types/pvc.go:148]: **Binding** can be used as adjective and noun. And this is misleading. 

I think **HasUnboundPvcs** would be easier to understand.

File: pkg/virt-controller/watch/vm.go [hunk 2]
```diff
@@ -436,7 +376,7 @@ func (c *VMController) handleDataVolumes(vm *virtv1.VirtualMachine, dataVolumes
 		}
 		if !exists {
 			// Don't create DV if PVC already exists
-			pvc, err := c.getPersistentVolumeClaimFromCache(vm.Namespace, template.Name)
+			pvc, err := typesutil.GetPersistentVolumeClaimFromCache(vm.Namespace, template.Name, c.pvcInformer)
```
Comment (brybacki) [inline pkg/virt-controller/watch/vm.go:389]: I do not like the name 	**typesutil** "kubevirt.io/kubevirt/pkg/storage/types".

What do you think about **storage**, or **storagetypes**

File: pkg/storage/types/dv.go [hunk 3]
```diff
@@ -81,7 +108,7 @@ func GetCloneSourceWithInformers(vm *virtv1.VirtualMachine, dvSpec *cdiv1.DataVo
 	return cloneSource, nil
 }
 
-func GetCloneSource(ctx context.Context, client kubecli.KubevirtClient, vm *virtv1.VirtualMachine, dvSpec *cdiv1.DataVolumeSpec) (*CloneSource, error) {
+func GetCloneSource(ctx context.Context, client kubecli.KubevirtClient, namespace string, dvSpec *cdiv1.DataVolumeSpec) (*CloneSource, error) {
```
Comment (brybacki) [inline pkg/storage/types/dv.go:111]: :+1:

File: pkg/virt-controller/watch/vmi.go [hunk 4]
```diff
@@ -327,7 +308,7 @@ func (c *VMIController) execute(key string) error {
 	}
 
 	// Get all dataVolumes associated with this vmi
-	dataVolumes, err := c.listMatchingDataVolumes(vmi)
+	dataVolumes, err := kubevirttypes.ListDataVolumesFromVolumes(vmi.Namespace, vmi.Spec.Volumes, c.dataVolumeInformer, c.pvcInformer)
```
Comment (brybacki) [inline pkg/virt-controller/watch/vmi.go:308]: here again, I think storagetypes would be better

File: pkg/storage/types/pvc.go [hunk 5]
```diff
@@ -195,3 +198,36 @@ func VolumeReadyToAttachToNode(namespace string, volume virtv1.Volume, dataVolum
 	}
 	return ready, wffc, nil
 }
+
+func GeneratePVC(size *resource.Quantity, claimName, namespace, storageClass, accessMode string, blockVolume bool) *k8sv1.PersistentVolumeClaim {
```
Comment (brybacki) [inline pkg/storage/types/pvc.go:202]: I think the word used in kubevirt for creating the `spec` is `render` not `generate`.

**Review by brybacki — COMMENTED** [review]
Very nice change. Separates storage domain from networking and virt code. 

Had just a few minor comments.

File: pkg/storage/types/pvc.go (hunk 5 above)
Comment (brybacki) [reply pkg/storage/types/pvc.go:202]: as in RenderPod, etc...

File: pkg/storage/types/pvc.go [hunk 6]
```diff
@@ -155,3 +157,45 @@ func GetSizeIncludingFSOverhead(size *resource.Quantity, storageClass *string, v
 	cdiFSOverhead := GetFilesystemOverhead(volumeMode, storageClass, cdiConfig)
 	return GetSizeIncludingGivenOverhead(size, cdiFSOverhead)
 }
+
+func GetPersistentVolumeClaimFromCache(namespace, name string, pvcInformer cache.SharedIndexInformer) (*k8sv1.PersistentVolumeClaim, error) {
```
[the remark this answers is not in the document]
Comment (ShellyKa13) [reply pkg/storage/types/pvc.go:130]: I'm passing to the function the pvcinformer and dataVolumeInformer of the controller which are SharedIndexInformer : https://github.com/kubevirt/kubevirt/blob/330c8236dc89ed58ea6e860eb4c9eca1ae2b5184/pkg/virt-controller/watch/vm.go#L169-L170

[the remark this answers is not in the document]
Comment (brybacki) [reply pkg/storage/types/pvc.go:130]: Yes, but SharedIndexInformer satisfies the SharedInformer, And I think it is always good to select minimal interface. But this is very minor comment.

[the remark this answers is not in the document]
Comment (ShellyKa13) [reply pkg/storage/types/pvc.go:130]: ahh ok I didnt know that, if it works its fine by me to change :)

**Comment by brybacki** [comment]
@ShellyKa13 you need to rebase, and then we need to finish the review and contact some approvers, otherwise it will need another rebase.

**Comment by ShellyKa13** [comment]
@brybacki Yeah I was doing that and tried to run the unit tests to make sure the rebase didnt break anything and then I was hit with this issue: https://github.com/kubevirt/kubevirt/pull/8403

**Comment by brybacki** [comment]
> @brybacki Yeah I was doing that and tried to run the unit tests to make sure the rebase didnt break anything and then I was hit with this issue: #8403

What do you mean? The link points to a merged PR

**Comment by ShellyKa13** [comment]
@brybacki If you read in the link the comments below youll see they write about an issue with unit-tests that that PR introduced.
Anyways there is a temp fix which I used and everything passed. so you can review again

**Pull request merged**